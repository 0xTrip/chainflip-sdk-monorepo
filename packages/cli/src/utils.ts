import { createInterface } from 'node:readline/promises';
import yargs from 'yargs/yargs';
import { Assets, ChainflipNetworks } from '@/shared/enums';
import { chainflipNetwork } from '@/shared/parsers';
import { ChainflipNetwork } from './enums';

export const askForPrivateKey = async () => {
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  try {
    return await rl.question("Please enter your wallet's private key: ");
  } finally {
    rl.close();
  }
};

type GetEthNetworkOptions =
  | { chainflipNetwork: 'localnet'; ethNetwork: string }
  | { chainflipNetwork: ChainflipNetwork };

export function getEthNetwork(opts: GetEthNetworkOptions) {
  if (opts.chainflipNetwork === 'localnet') return opts.ethNetwork;
  if (opts.chainflipNetwork === ChainflipNetworks.mainnet) return 'mainnet';
  return 'goerli';
}
const networks = [...Object.values(chainflipNetwork.enum), 'localnet'];
export const parseArgs = (args: string[]) =>
  yargs(args)
    .scriptName('chainflip-cli')
    .usage('$0 <cmd> [args]')
    .command('swap', '', (y) => {
      y.option('src-asset', {
        choices: Object.values(Assets),
        // demandOption: true,
        describe: 'The asset to swap from',
      })
        .option('dest-asset', {
          choices: Object.values(Assets),
          demandOption: true,
          describe: 'The asset to swap to',
        })
        .option('chainflip-network', {
          choices: networks,
          describe: 'The Chainflip network to execute the swap on',
          default: ChainflipNetworks.sisyphos,
        })
        .option('amount', {
          type: 'string',
          demandOption: true,
          describe: 'The amount to swap',
        })
        .option('dest-address', {
          type: 'string',
          demandOption: true,
          describe: 'The address to send the swapped assets to',
        })
        .option('wallet-private-key', {
          type: 'string',
          describe: 'The private key of the wallet to use',
        })
        .option('src-token-contract-address', {
          type: 'string',
          describe:
            'The contract address of the token to swap from when `chainflip-network` is `localnet`',
        })
        .option('vault-contract-address', {
          type: 'string',
          describe:
            'The contract address of the vault when `chainflip-network` is `localnet`',
        })
        .option('eth-network', {
          type: 'string',
          describe:
            'The eth network URL to use when `chainflip-network` is `localnet`',
        });
    })
    .command('fund-state-chain-account', '', (y) => {
      y.option('account-id', {
        type: 'string',
        demandOption: true,
        describe: 'The account ID for the validator to be funded',
      })
        .option('chainflip-network', {
          choices: networks,
          describe: 'The Chainflip network to execute the swap on',
          default: ChainflipNetworks.sisyphos,
        })
        .option('amount', {
          type: 'string',
          demandOption: true,
          describe: 'The amount in Flipperino to fund',
        })
        .option('wallet-private-key', {
          type: 'string',
          describe: 'The private key of the wallet to use',
        })
        .option('state-chain-manager-contract-address', {
          type: 'string',
          describe:
            'The contract address of the state chain manager when `chainflip-network` is `localnet`',
        })
        .option('flip-token-contract-address', {
          type: 'string',
          describe:
            'The contract address for the FLIP token when `chainflip-network` is `localnet`',
        })
        .option('eth-network', {
          type: 'string',
          describe:
            'The eth network URL to use when `chainflip-network` is `localnet`',
        });
    })
    .help()
    .parseSync();
