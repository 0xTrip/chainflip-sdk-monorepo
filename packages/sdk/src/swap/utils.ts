export const unreachable = (x: never, message: string): never => {
  throw new Error(`${message}: ${x}`);
};
