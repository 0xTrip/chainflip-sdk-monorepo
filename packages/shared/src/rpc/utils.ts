import { UncheckedAssetAndChain, assertIsValidAssetAndChain } from '../enums';
import { bigintMax } from '../functions';
import { ChainAssetMap, Environment } from './index';

export const readAssetValue = <T>(
  value: ChainAssetMap<T>,
  asset: UncheckedAssetAndChain,
): T => {
  assertIsValidAssetAndChain(asset);
  const chainMinimums = value[asset.chain];
  return chainMinimums[asset.asset as keyof typeof chainMinimums];
};

export const readMinimumSwapAmount = (
  env: Pick<Environment, 'swapping' | 'ingressEgress'>,
  asset: UncheckedAssetAndChain,
): bigint => {
  const minimumSwapAmount = readAssetValue(
    env.swapping.minimumSwapAmounts,
    asset,
  );
  const minimumDepositAmount = readAssetValue(
    env.ingressEgress.minimumDepositAmounts,
    asset,
  );
  return bigintMax(minimumSwapAmount, minimumDepositAmount);
};

type Result = { success: true } | { success: false; reason: string };

export const validateSwapAmount = (
  env: Environment,
  asset: UncheckedAssetAndChain,
  amount: bigint,
): Result => {
  const minimumAmount = readMinimumSwapAmount(env, asset);

  if (amount < minimumAmount) {
    return {
      success: false,
      reason: `expected amount is below minimum swap amount (${minimumAmount})`,
    };
  }

  const maxAmount = readAssetValue(env.swapping.maximumSwapAmounts, asset);

  if (maxAmount != null && amount > maxAmount) {
    return {
      success: false,
      reason: `expected amount is above maximum swap amount (${maxAmount})`,
    };
  }

  return { success: true };
};
