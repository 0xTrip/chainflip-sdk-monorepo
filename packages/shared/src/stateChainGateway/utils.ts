import { StateChainGateway__factory } from '../abis';
import { getStateChainGatewayContractAddress } from '../contracts';
import type { FundingNetworkOptions } from './index';

export const getStateChainGateway = (options: FundingNetworkOptions) => {
  const stateChainGatewayContractAddress =
    options.network === 'localnet'
      ? options.stateChainGatewayContractAddress
      : getStateChainGatewayContractAddress(options.network);

  return StateChainGateway__factory.connect(
    stateChainGatewayContractAddress,
    options.signer,
  );
};
