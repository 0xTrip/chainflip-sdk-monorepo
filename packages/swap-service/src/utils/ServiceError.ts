export default class ServiceError extends Error {
  static badRequest(message: string): ServiceError {
    return new ServiceError(message, 400);
  }

  static notFound(message = 'resource not found'): ServiceError {
    return new ServiceError(message, 404);
  }

  constructor(message: string, readonly code: number) {
    super(message);

    Error.captureStackTrace(this, ServiceError);
  }
}
