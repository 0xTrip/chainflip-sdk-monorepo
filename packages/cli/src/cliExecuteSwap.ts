import { getDefaultProvider, providers, Wallet } from 'ethers';
import { createInterface } from 'readline/promises';
import { z } from 'zod';
import { ChainId } from '@/sdk/swap/consts';
import { executeSwap, ExecuteSwapParams } from '@/sdk/swap/vault';
import {
  chainflipNetwork,
  SupportedAsset,
  supportedAsset,
} from '@/shared/enums';
import { numericString, hexString } from '@/shared/parsers';

const assetToNetworkMap: Record<SupportedAsset, ChainId> = {
  ETH: ChainId.Ethereum,
  FLIP: ChainId.Ethereum,
  USDC: ChainId.Ethereum,
  BTC: ChainId.Bitcoin,
  DOT: ChainId.Polkadot,
};

const argsSchema = z
  .object({
    srcToken: supportedAsset.optional(),
    destToken: supportedAsset,
    amount: z.union([numericString, hexString]),
    destAddress: z.string(),
    chainflipNetwork,
    _: z.tuple([z.literal('swap')]),
  })
  .transform(({ destToken, srcToken, _, ...rest }) => {
    const destChainId = assetToNetworkMap[destToken];

    return {
      ...rest,
      destChainId,
      destTokenSymbol: destToken,
      ...(srcToken && { srcTokenSymbol: srcToken }),
    };
  });

export default async function cliExecuteSwap(args: unknown) {
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  try {
    const { chainflipNetwork: cfNetwork, ...validatedArgs } =
      argsSchema.parse(args);

    const privateKey = await rl.question(
      "Please enter your wallet's private key: ",
    );

    const ethNetwork = cfNetwork === 'mainnet' ? 'mainnet' : 'goerli';

    const wallet = new Wallet(privateKey).connect(
      process.env.ALCHEMY_KEY
        ? new providers.AlchemyProvider(ethNetwork, process.env.ALCHEMY_KEY)
        : getDefaultProvider(ethNetwork),
    );

    const txHash = await executeSwap(
      validatedArgs as ExecuteSwapParams,
      cfNetwork,
      wallet,
    );

    console.log(`Swap executed. Transaction hash: ${txHash}`);
  } finally {
    rl.close();
  }
}
