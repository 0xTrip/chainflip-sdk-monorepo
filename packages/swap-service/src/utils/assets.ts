import { hexToU8a, isHex } from '@polkadot/util';
import { decodeAddress, encodeAddress } from '@polkadot/util-crypto';
import * as ethers from 'ethers';
import { z } from 'zod';
import { SupportedAsset, supportedAsset } from '@/shared/assets';
import ServiceError from './ServiceError';

export const stateChainAsset = z
  .enum(['Usdc', 'Flip', 'Dot', 'Eth'])
  .transform((val) => val.toUpperCase() as SupportedAsset);

export const validateAddress = (
  egressAsset: SupportedAsset,
  address: string,
): void => {
  if (
    (egressAsset === 'ETH' ||
      egressAsset === 'FLIP' ||
      egressAsset === 'USDC') &&
    ethers.isAddress(address)
  ) {
    return;
  }

  if (egressAsset === 'DOT') {
    try {
      encodeAddress(
        isHex(address) ? hexToU8a(address) : decodeAddress(address),
      );

      return;
    } catch {
      // pass
    }
  }

  throw ServiceError.badRequest('provided address is not valid');
};

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
