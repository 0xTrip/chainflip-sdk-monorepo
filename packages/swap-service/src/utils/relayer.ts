import { z } from 'zod';
import { supportedAsset, SupportedAsset } from './assets';
import { memoize } from './function';
import { bareHexString, hexString, numericString } from './parsers';
import RpcClient from './RpcClient';
import { transformAsset } from './string';

type NewSwapIntent = {
  ingressAsset: SupportedAsset;
  egressAsset: SupportedAsset;
  egressAddress: string;
};

const address = z.union([hexString, bareHexString.transform((v) => `0x${v}`)]);

const requestValidators = {
  newSwapIngressAddress: z
    .tuple([
      supportedAsset.transform(transformAsset),
      supportedAsset.transform(transformAsset),
      numericString.or(hexString),
      z.number(),
    ])
    .transform(([a, b, c, d]) => [a, b, c, d]),
};

const validators = {
  newSwapIngressAddress: address,
};

const initializeClient = memoize(async () => {
  const rpcClient = await new RpcClient(
    process.env.RPC_RELAYER_WSS_URL as string,
    requestValidators,
    validators,
    'relayer',
  ).connect();

  return rpcClient;
});

export const submitSwapToRelayer = async (
  swapIntent: NewSwapIntent,
): Promise<string> => {
  const client = await initializeClient();

  const ingressAddress = await client.sendRequest(
    'newSwapIngressAddress',
    swapIntent.ingressAsset,
    swapIntent.egressAsset,
    swapIntent.egressAddress,
    0,
  );

  return ingressAddress;
};
