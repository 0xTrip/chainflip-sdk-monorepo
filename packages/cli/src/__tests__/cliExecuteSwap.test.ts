import { schema } from '../cliExecuteSwap';
import { parseArgs } from '../utils';

const localnet = `swap
  --wallet-private-key 0x2
  --chainflip-network localnet
  --dest-token USDC
  --amount 1000000000
  --dest-address 0x0
  --src-token-contract-address 0x0
  --vault-contract-address 0x0
  --eth-network test`;

describe('schema', () => {
  it.each([localnet, localnet.replace('localnet', 'sisyphos')])(
    'properly parses the arguments',
    (args) => {
      expect(schema.parse(parseArgs(args.split(/\s+/)))).toMatchSnapshot();
    },
  );
});
