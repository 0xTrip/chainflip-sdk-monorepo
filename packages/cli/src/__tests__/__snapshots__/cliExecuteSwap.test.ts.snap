// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema properly parses the arguments 1`] = `
{
  "amount": "1000000000",
  "chainflipNetwork": "localnet",
  "destAddress": "0x0",
  "destChainId": 1,
  "destTokenSymbol": "USDC",
  "ethNetwork": "test",
  "srcTokenContractAddress": "0x0",
  "vaultContractAddress": "0x0",
  "walletPrivateKey": "0x2",
}
`;

exports[`schema properly parses the arguments 2`] = `
{
  "amount": "1000000000",
  "chainflipNetwork": "sisyphos",
  "destAddress": "0x0",
  "destChainId": 1,
  "destTokenSymbol": "USDC",
  "walletPrivateKey": "0x2",
}
`;
